src-ghc/Pact/Types/Crypto.hs

```diff
@@ -217,61 +183,87 @@ instance ConvertBS Ed25519.Signature where
-  toBS = ECDSA.exportPrivate
-  fromBS s = maybeToEither ("Invalid ECDSA Private Key: " ++ show (toB16Text s))
-             (ECDSA.importPrivate s)
-instance ConvertBS (ECDSA.Signature) where
-  toBS = ECDSA.exportSignature
-  fromBS s = maybeToEither ("Invalid ECDSA Signature: " ++ show (toB16Text s))
-             (ECDSA.importSignature s)
-
-
-
+instance Scheme (SPPKScheme 'WebAuthn) where
+
+  _valid _ msg (PubBS pubBS) (SigBS sigBS) =
+  -- _valid _ msg UserSig{..} Signer{..} =
+    case runVerification of
+      Left e -> error e
+      Right () -> True
+    where
+      -- Verifying a WebAuthn signature requires that we know the payload
+      -- signed by the WebAuthn keys on the client device. This payload is
+      -- a combination of the Challenge (in our case, the PactHash of a
+      -- transaction), a JSON object of "client metadata", and data about
+      -- the authenticator hardware.
+      -- We require this data to be part of our WebAuthn signature so that
+      -- we can reconstitute the payload that was signed on the browser.
+      runVerification :: Either String ()
+      runVerification = do
+
+        -- Decode our WebAuthn signature object from a `UserSig` string.
+        WebAuthnSignature
+          { authenticatorData
+          , signature
+          , clientDataJSON } <- A.eitherDecode (BSL.fromStrict sigBS)
+        -- Decode the signer's public key.
+        publicKeyBytes <- return pubBS -- Base16.decode (T.encodeUtf8 pubBS)
+        publicKey <- first show $ Serialise.deserialiseOrFail @WA.CosePublicKey (BSL.fromStrict publicKeyBytes)
+
+        -- Recover the signature, clientData, and authData bytestrings.
+        -- Lenient decoding is not a security risk because security comes
+        -- from valitating the signature against the payload. Invalid
+        -- bytestring encodings (ones that pass lenient decoding but fail
+        -- strict decoding) will parse, but fail signature validation.
+        let
+          sig = Base64.decodeLenient (T.encodeUtf8 signature)
+          clientData = Base64URL.decodeLenient (T.encodeUtf8 clientDataJSON)
+          authData = Base64.decodeLenient (T.encodeUtf8 authenticatorData)
+
+        -- Reconstitute the payload signed by the WebAuthn client.
+        clientDataDigest <- Base16.decode $ BS.pack (show (H.hashWith H.SHA256 clientData))
+        let
+          payload = authData <> clientDataDigest
+
+        -- Check the signature's validity.
+        first T.unpack $ WAVerify.verify publicKey payload sig
+
+        -- Extract the original challenge from client data.
+        ClientDataJSON { challenge } <- A.eitherDecode (BSL.fromStrict clientData)
+
+        -- Check that the input `PactHash` matches the hash of the transaction
+        -- that was signed by WebAuthn keys.
+        let pactHashText :: T.Text =
+              -- PactHash.hashToText $ PactHash.toUntypedHash msg
+              PactHash.hashToText msg -- TODO: correct?
```

Comment:
I think this is correct (b64 unpadded)

Reply:
It should be unpadded, URL-safe base64 - just to make sure we're on the same page.

Reply:
:+1: `hashToText` does an unpadded Base64URL encode.